nerfstudio/cameras/cameras.py

```diff
@@ -723,56 +696,61 @@ def _generate_rays_from_coords(
+            directions_stack[..., 1][mask] = torch.masked_select(coord_stack[..., 1] * sin_theta / theta, mask).float()
+            directions_stack[..., 2][mask] = -torch.masked_select(torch.cos(theta), mask).float()
 
-            # local axes of ODS camera
-            ods_x_axis = torch.tensor([1, 0, 0]).to("cuda")
-            ods_z_axis = torch.tensor([0, 0, -1]).to("cuda")
+        if CameraType.EQUIRECTANGULAR.value in cam_types:
+            mask = (self.camera_type[true_indices] == CameraType.EQUIRECTANGULAR.value).squeeze(-1)  # (num_rays)
+            mask = torch.stack([mask, mask, mask], dim=0)
 
-            # circle of ODS ray origins
-            ods_origins_circle = (
-                ods_cam_position
-                + (vr_ipd / 2.0) * (ods_x_axis.repeat(c2w.shape[1], 1)) * (torch.cos(ods_theta))[:, None]
-                + (vr_ipd / 2.0) * (ods_z_axis.repeat(c2w.shape[1], 1)) * (torch.sin(ods_theta))[:, None]
-            )
+            # For equirect, fx = fy = height = width/2
+            # Then coord[..., 0] goes from -1 to 1 and coord[..., 1] goes from -1/2 to 1/2
+            theta = -torch.pi * coord_stack[..., 0]  # minus sign for right-handed
+            phi = torch.pi * (0.5 - coord_stack[..., 1])
+            # use spherical in local camera coordinates (+y up, x=0 and z<0 is theta=0)
+            directions_stack[..., 0][mask] = torch.masked_select(-torch.sin(theta) * torch.sin(phi), mask).float()
+            directions_stack[..., 1][mask] = torch.masked_select(torch.cos(phi), mask).float()
+            directions_stack[..., 2][mask] = torch.masked_select(-torch.cos(theta) * torch.sin(phi), mask).float()
 
-            # rotate origins to match the camera rotation
-            for i in range(ods_origins_circle.shape[0]):
-                ods_origins_circle[i] = rotation[0][0] @ ods_origins_circle[i] + ods_cam_position[0]
-            ods_origins_circle = ods_origins_circle.unsqueeze(0).repeat(c2w.shape[0], 1, 1)
+        if CameraType.OMNIDIRECTIONALSTEREO_L.value in cam_types:
+            ods_origins_circle, directions_stack = omni_directional_stereo_camera("left")
+            # assign final camera origins
+            c2w[..., :3, 3] = ods_origins_circle
 
+        if CameraType.OMNIDIRECTIONALSTEREO_R.value in cam_types:
+            ods_origins_circle, directions_stack = omni_directional_stereo_camera("right")
             # assign final camera origins
             c2w[..., :3, 3] = ods_origins_circle
 
-        for value in cam_types:
-            if value not in [
-                CameraType.PERSPECTIVE.value,
-                CameraType.FISHEYE.value,
-                CameraType.EQUIRECTANGULAR.value,
-                CameraType.OMNIDIRECTIONALSTEREO_L.value,
-                CameraType.OMNIDIRECTIONALSTEREO_R.value,
-            ]:
-                raise ValueError(f"Camera type {value} not supported.")
+        for cam_type in cam_types:
```

Comment:
Sorry I should have been more specific. The logic above should exist in each if condition, ie:

```
for cam_type in cam_types:
    if cam_type == CameraType.PERSPECTIVE.value:
        mask = (self.camera_type[true_indices] == CameraType.PERSPECTIVE.value).squeeze(-1)  # (num_rays)
        mask = torch.stack([mask, mask, mask], dim=0)
        directions_stack[..., 0][mask] = torch.masked_select(coord_stack[..., 0], mask).float()
        directions_stack[..., 1][mask] = torch.masked_select(coord_stack[..., 1], mask).float()
        directions_stack[..., 2][mask] = -1.0
    elif cam_type == CameraType.EQUIRECTANGULAR.value:
        # Fill in
    # Add other cases
    else:
        raise ValueError(f"Camera type {cam_type} not supported.")
```